### PR TITLE
Add implementation for hint on abs_value (math.cairo)

### PR DESCRIPTION
**Pull request by fmoletta**
- Add is_positive on hint_utils
- Add as_int on math_utils
- Add unit test and integration test with abs_value function
Description of the pull request changes and motivation.

## Checklist
- [ ] Linked to Github Issue
- [x] Unit tests added
- [x] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

File: src/math_utils.rs [hunk 1]
```diff
@@ -3,6 +3,15 @@ use num_bigint::BigInt;
 use num_integer::Integer;
 use num_traits::{abs, FromPrimitive};
 
+/// Returns the lift of the given field element, val, as an integer in the range (-prime/2, prime/2).
+pub fn as_int(val: &BigInt, prime: &BigInt) -> BigInt {
```
Comment (Oppen) [inline src/math_utils.rs:7]: Why is it called `as_int`?

File: cairo_programs/abs_value_array.cairo [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    tempvar array_length = 10
+    let (array_a : felt*) = alloc()
+
+    let (array_b : felt*) = alloc()
+    fill_array(array_a, -7, -3, array_length, 0)
+    fill_array(array_b, 7, 3, array_length, 0)
+    let result : felt = compare_abs_arrays(array_a, array_b, array_length, 0)
+    assert result = TRUE
+    return ()
+end
```
Comment (Oppen) [inline cairo_programs/abs_value_array.cairo:38]: I don't find the hint here 🤔

File: src/vm/hints/hint_utils.rs [hunk 3]
```diff
@@ -275,3 +276,76 @@ pub fn assert_le_felt(
...
+                    } else {
+                        return Err(VirtualMachineError::NoRangeCheckBuiltin);
+                    };
+                    //Main logic (assert a is positive)
+                    let mut result = bigint!(0);
+                    let int_value = as_int(value, &vm.prime);
+                    if int_value.abs() > range_check_builtin._bound {
+                        return Err(VirtualMachineError::ValueOutsideValidRange(int_value));
+                    }
+                    if int_value > bigint!(0) {
```
Comment (Oppen) [inline src/vm/hints/hint_utils.rs:333]:
```suggestion
                    if int_value.is_positive() {
```

File: src/vm/hints/hint_utils.rs [hunk 4]
```diff
@@ -275,3 +276,76 @@ pub fn assert_le_felt(
...
+                        result = bigint!(1);
+                    }
+                    return match vm
+                        .memory
+                        .insert(&is_positive_addr, &MaybeRelocatable::from(result))
+                    {
+                        Ok(_) => Ok(()),
+                        Err(memory_error) => Err(VirtualMachineError::MemoryError(memory_error)),
+                    };
```
Comment (Oppen) [inline src/vm/hints/hint_utils.rs:342]: Not sure if it'd work here, but I suggest trying:
```suggestion
                        .insert(&is_positive_addr, &MaybeRelocatable::from(result))
                        .map_err(VirtualMachineError::MemoryError);
```

File: src/vm/hints/hint_utils.rs [hunk 5]
```diff
@@ -275,3 +276,76 @@ pub fn assert_le_felt(
...
+                }
+            }
+            Err(VirtualMachineError::NoRangeCheckBuiltin)
+        }
+        (Err(memory_error), _) => Err(VirtualMachineError::MemoryError(memory_error)),
+        (_, Err(memory_error)) => Err(VirtualMachineError::MemoryError(memory_error)),
```
Comment (Oppen) [inline src/vm/hints/hint_utils.rs:348]:
```suggestion
        (Err(memory_error), _)|(_, Err(memory_error)) => Err(VirtualMachineError::MemoryError(memory_error)),
```

File: src/math_utils.rs (hunk 1 above)
Comment (fmoletta) [reply src/math_utils.rs:7]: Thats the name they gave it in math_utils.py in the common library

Comment (fmoletta) [reply src/math_utils.rs:7]: This function is used by various hints, so its better to keep the same name

File: src/vm/hints/hint_utils.rs (hunk 4 above)
Comment (fmoletta) [reply src/vm/hints/hint_utils.rs]: It works!

**Review by Juan-M-V — APPROVED** [review]

File: src/math_utils.rs (hunk 1 above)
Comment (Oppen) [reply src/math_utils.rs:7]: Works for me.

File: src/vm/hints/hint_utils.rs [hunk 6]
```diff
@@ -277,6 +279,135 @@ pub fn assert_le_felt(
...
+    ) {
+        (a_addr, b_addr)
+    } else {
+        return Err(VirtualMachineError::FailedToGetIds);
+    };
+    //Check that the ids are in memory
+    match (vm.memory.get(&a_addr), vm.memory.get(&b_addr)) {
+        (Ok(Some(maybe_rel_a)), Ok(Some(maybe_rel_b))) => match (maybe_rel_a, maybe_rel_b) {
+            (MaybeRelocatable::Int(ref a), MaybeRelocatable::Int(ref b)) => {
+                if (a - b).mod_floor(&vm.prime) == bigint!(0) {
```
Comment (Oppen) [inline src/vm/hints/hint_utils.rs:318]:
```suggestion
                if (a - b).is_multiple_of(&vm.prime) {
```

File: src/vm/hints/hint_utils.rs [hunk 7]
```diff
@@ -277,6 +279,135 @@ pub fn assert_le_felt(
...
+            let range_check_builtin = if let Some(range_check_builtin) =
+                builtin.as_any().downcast_ref::<RangeCheckBuiltinRunner>()
+            {
+                range_check_builtin
+            } else {
+                return Err(VirtualMachineError::NoRangeCheckBuiltin);
+            };
+
+            // assert 0 <= ids.a % PRIME < range_check_builtin.bound
+            if a.mod_floor(&vm.prime).is_positive()
```
Comment (Oppen) [inline src/vm/hints/hint_utils.rs:400]: `.mod_floor` is always positive if `vm.prime` is positive.

Comment (fmoletta) [reply src/vm/hints/hint_utils.rs]: Fixed!

File: src/vm/hints/hint_utils.rs [hunk 8]
```diff
@@ -407,3 +406,73 @@ pub fn assert_nn(
...
+                    } else {
+                        return Err(VirtualMachineError::NoRangeCheckBuiltin);
+                    };
+                    //Main logic (assert a is positive)
+                    let mut result = bigint!(0);
+                    let int_value = as_int(value, &vm.prime);
+                    if int_value.abs() > range_check_builtin._bound {
+                        return Err(VirtualMachineError::ValueOutsideValidRange(int_value));
+                    }
+                    if int_value.is_positive() {
```
[the remark this answers is not in the document]
Comment (Oppen) [reply src/vm/hints/hint_utils.rs:462]: Good idea. It would save an object creation.

[the remark this answers is not in the document]
Comment (fmoletta) [reply src/vm/hints/hint_utils.rs:462]: Good idea + 1

File: src/vm/hints/hint_utils.rs [hunk 9]
```diff
@@ -162,13 +162,14 @@ pub fn is_nn_out_of_range(
                         return Err(VirtualMachineError::NoRangeCheckBuiltin);
                     };
                     //Main logic (assert a is not negative and within the expected range)
-                    let mut value = bigint!(1);
-                    if (-a.clone() - bigint!(1)).mod_floor(&vm.prime) >= bigint!(0)
+                    let value = if (-a.clone() - bigint!(1)).mod_floor(&vm.prime) >= bigint!(0)
```
Comment (Oppen) [inline src/vm/hints/hint_utils.rs:165]: `.mod_floor` is always non-negative. There's also no need for cloning `a`.
```suggestion
                    let value = if (-a - &bigint!(1)).mod_floor(&vm.prime)
```

Comment (fmoletta) [reply src/vm/hints/hint_utils.rs]: Done!

File: src/math_utils.rs [hunk 10]
```diff
@@ -3,6 +3,15 @@ use num_bigint::BigInt;
...
+    if val < &prime.div_floor(&bigint!(2)) {
```
[the remark this answers is not in the document]
Comment (fmoletta) [reply src/math_utils.rs]: Done!

**Review by Oppen — APPROVED** [review]

**Pull request merged**